### PR TITLE
SoundWire: improve bus reset

### PR DESCRIPTION
**Pull request by plbossart**
Tentative PR to check CI results on multiple generations of hardware. Tested successfully with MODEL=TGLU_SKU0A32_SDCA

### CODE REVIEW

File: drivers/soundwire/intel_bus_common.c [hunk 1]
```diff
@@ -125,12 +129,23 @@ int intel_start_bus_after_reset(struct sdw_intel *sdw)
...
+			dev_err(dev, "unable to exit bus reset sequence during resume\n");
+			return ret;
+		}
+
 		ret = sdw_cdns_enable_interrupt(cdns, true);
 		if (ret < 0) {
 			dev_err(dev, "cannot enable interrupts during resume\n");
 			return ret;
 		}
-
```
Comment (ranj063) [inline drivers/soundwire/intel_bus_common.c]: spurios change

File: drivers/soundwire/intel_bus_common.c [hunk 2]
```diff
@@ -115,31 +106,35 @@ int intel_start_bus_after_reset(struct sdw_intel *sdw)
 	ret = sdw_cdns_clock_restart(cdns, !clock_stop0);
 	if (ret < 0) {
 		dev_err(dev, "unable to restart clock during resume\n");
-		goto err_interrupt;
+		if (!clock_stop0)
+			sdw_cdns_enable_interrupt(cdns, false);
```
Comment (plbossart) [inline drivers/soundwire/intel_bus_common.c:118]: note to self: missing return ret;

File: drivers/soundwire/intel_bus_common.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/intel_bus_common.c]: yes indeed.

File: drivers/soundwire/intel_bus_common.c (hunk 2 above)
Comment (plbossart) [reply drivers/soundwire/intel_bus_common.c:118]: fixed

File: drivers/soundwire/intel_bus_common.c (hunk 1 above)
Comment (plbossart) [reply drivers/soundwire/intel_bus_common.c]: fixed

**Comment by rfvirgil** [comment]
Tested on TigerLake UpXtreme + 4 x CS35L56 (EIGHT-C/AIC1.0).When exiting clock stop there's some initial bus activity which is too short for the analyzer to decode, but I assume it is PINGs. Then a full bus-reset and CS35L56 audio is working.
This isn't a stress test, but I've repeated this manually several times and I see a bus-reset on every exit from clock-stop.

image

**Comment by plbossart** [comment]
Thanks @rfvirgil for testing, I'll do more testing on my end when our CI is back in gear. Happy Friday!

**Comment by plbossart** [comment]
SOFCI TEST

File: drivers/soundwire/cadence_master.c [hunk 3]
```diff
@@ -283,6 +283,29 @@ static int cdns_config_update(struct sdw_cdns *cdns)
...
 }
 
+/**
+ * sdw_cdns_config_update() - Update configurations
+ * @cdns: Cadence instance
+ */
+void sdw_cdns_config_update(struct sdw_cdns *cdns)
+{
+	/* commit changes */
+	cdns_writel(cdns, CDNS_MCP_CONFIG_UPDATE, CDNS_MCP_CONFIG_UPDATE_BIT);
```
Comment (RanderWang) [inline drivers/soundwire/cadence_master.c:293]: do we need to use cdns_updatel ?

**Review by RanderWang — COMMENTED** [review]
LGTM

File: drivers/soundwire/cadence_master.c (hunk 3 above)
Comment (plbossart) [reply drivers/soundwire/cadence_master.c:293]: It's not necessary @RanderWang 

CONFIG_UPDATE only has BIT(0) defined, bits 1..31 are reserved. So it's useless to have a read-modify-write.

The only times where a read is necessary is *after* a write, when making sure that the bit was indeed cleared by hardware.

We can actually clean-up all other usages of CONFIG_UPDATE, I'll add a patch for this to be consistent. 

Thanks for the feedback.

**Review by bardliao — DISMISSED** [review]

**Comment by plbossart** [comment]
v2 added, the main change is the addition of a wait in exit_reset

[v2.diff.txt](https://github.com/thesofproject/linux/files/11193322/v2.diff.txt)

````diff
diff --git a/drivers/soundwire/cadence_master.c b/drivers/soundwire/cadence_master.c
index 9ca600f1fa19..eb1b87ad1681 100644
--- a/drivers/soundwire/cadence_master.c
+++ b/drivers/soundwire/cadence_master.c
@@ -456,9 +456,9 @@ static int cdns_parity_error_injection(void *data, u64 value)
 			CDNS_IP_MCP_CMDCTRL_INSERT_PARITY_ERR);
 
 	/* commit changes */
-	cdns_updatel(cdns, CDNS_MCP_CONFIG_UPDATE,
-		     CDNS_MCP_CONFIG_UPDATE_BIT,
-		     CDNS_MCP_CONFIG_UPDATE_BIT);
+	ret = cdns_clear_bit(cdns, CDNS_MCP_CONFIG_UPDATE, CDNS_MCP_CONFIG_UPDATE_BIT);
+	if (ret < 0)
+		goto unlock;
 
 	/* do a broadcast dummy read to avoid bus clashes */
 	ret = sdw_bread_no_pm_unlocked(&cdns->bus, 0xf, SDW_SCP_DEVID_0);
@@ -470,16 +470,17 @@ static int cdns_parity_error_injection(void *data, u64 value)
 			0);
 
 	/* commit changes */
-	cdns_updatel(cdns, CDNS_MCP_CONFIG_UPDATE,
-		     CDNS_MCP_CONFIG_UPDATE_BIT,
-		     CDNS_MCP_CONFIG_UPDATE_BIT);
-
-	/* Continue bus operation with parity error injection disabled */
-	mutex_unlock(&bus->bus_lock);
+	ret = cdns_clear_bit(cdns, CDNS_MCP_CONFIG_UPDATE, CDNS_MCP_CONFIG_UPDATE_BIT);
+	if (ret < 0)
+		goto unlock;
 
 	/* Userspace changed the hardware state behind the kernel's back */
 	add_taint(TAINT_USER, LOCKDEP_STILL_OK);
 
+unlock:
+	/* Continue bus operation with parity error injection disabled */
+	mutex_unlock(&bus->bus_lock);
+
 	/*
 	 * allow Master device to enter pm_runtime suspend. This may
 	 * also result in Slave devices suspending.
@@ -1139,13 +1140,7 @@ int sdw_cdns_exit_reset(struct sdw_cdns *cdns)
 		     CDNS_MCP_CONTROL_HW_RST);
 
 	/* commit changes */
-	cdns_updatel(cdns, CDNS_MCP_CONFIG_UPDATE,
-		     CDNS_MCP_CONFIG_UPDATE_BIT,
-		     CDNS_MCP_CONFIG_UPDATE_BIT);
-
-	/* don't wait here */
-	return 0;
-
+	return cdns_config_update(cdns);
 }
 EXPORT_SYMBOL(sdw_cdns_exit_reset);
 
diff --git a/drivers/soundwire/intel_bus_common.c b/drivers/soundwire/intel_bus_common.c
index be0c93106fc8..e5ac3cc7cb79 100644
--- a/drivers/soundwire/intel_bus_common.c
+++ b/drivers/soundwire/intel_bus_common.c
@@ -182,8 +182,7 @@ int intel_start_bus_after_clock_stop(struct sdw_intel *sdw)
 		return ret;
 	}
 
-	sdw_cdns_check_self_clearing_bits(cdns, "intel_resume_runtime no_quirks",
-					  true, INTEL_MASTER_RESET_ITERATIONS);
+	sdw_cdns_check_self_clearing_bits(cdns, __func__, true, INTEL_MASTER_RESET_ITERATIONS);
 
 	return 0;
 }
````

**Review by bardliao — APPROVED** [review]

**Review by RanderWang — APPROVED** [review]

**Pull request merged**